include/kernel.h

```diff
@@ -0,0 +1,529 @@
+ * @brief Try sending a mail to a mailbox. 
+ *
+ * This function tries to send a mail to a mailbox. If the
+ * mailbox is full, it returns false without blocking.
+ *
+ * @param[in] mbox The mailbox to send the mail to
+ * @param[in] mail The mail to send (must not be NULL)
+ *
+ * @return true if the mail was sent, false if the mailbox was full
+ */
+bool kmbox_try_send(kmbox_t *mbox, void *mail);
+
+/** 
+ * @brief Send a mail to a mailbox. 
+ *
+ * This function sends a mail to a mailbox. If the
+ * mailbox is full, it will block until another thread
+ * receive a pending mail, so that the mailbox has space
+ * for the new mail.
+ *
+ * @param[in] mbox The mailbox to send the mail to
+ * @param[in] mail The mail to send (must not be NULL)
+ *
+ */
+void kmbox_send(kmbox_t *mbox, void *mail);
+
+/** 
+ * @brief Try receiving a mail from a mailbox. 
+ *
+ * This function tries to receive a mail to a mailbox. If the
+ * mailbox is empty, it will return NULL without blocking.
+ *
+ * @param[in] mbox The mailbox to receive the mail from
+ *
+ * @return the received mail, or NULL if the mailbox was empty
+ *
+ */
+void* kmbox_try_recv(kmbox_t *mbox);
+
+/** 
+ * @brief Receive a mail from a mailbox. 
+ *
+ * This function receives a mail from a mailbox. If the
+ * mailbox is empty, it will block until a mail is available.
+ *
+ * @param[in] mbox The mailbox to receive the mail from
+ *
+ * @return the received mail
+ *
+ */
+void* kmbox_recv(kmbox_t *mbox);
+
+
+/**
+ * @brief Attach an event to a mailbox, so that a mail arrives when the event
+ *        is triggered.
+ *
+ * This functions connects a mailbox to the specified event, so that
+ * when the event is triggered, a mail is posted to the
+ * mailbox. The thread can then poll the mailbox to see when the
+ * event is arrived. The received mail will be the event itself.
+ *
+ * This function is useful for threads that need to react to multiple
+ * events and/or handle messages from other threads. If you just
+ * want to wait for a single interrupt, it is easier to simply
+ * call #thread_wait_event.
+ *
+ * Example:
+ *
+ *     kmbox_t *mbox = kmbox_new_stack(4);
+ *     kmbox_attach_event(mbox, &KEVENT_IRQ_SP);
+ *     kmbox_attach_event(mbox, &KEVENT_IRQ_DP);
+ *
+ *     void *mail;
+ *     while ((mail = kmbox_recv(mbox)))
+ *     {
+ *        if (mail == &KEVENT_IRQ_SP)
+ *        {
+ *            [...]
+ *        }
+ *        else if (mail == &KEVENT_IRQ_DP)
+ *        {
+ *            [...]
+ *        }
+ *     }
+ *
+ *     // Always detach mailboxes before freeing them!
+ *     kmbox_detach_event(mbox, &KEVENT_IRQ_SP);
+ *     kmbox_detach_event(mbox, &KEVENT_IRQ_DP);
+ *
+ *
+ * @note It is mandatory to detach all attached events before
+ *       destroying a mailbox.
+ *
+ * @note If the mailbox happens to be full when the event is triggered,
+ *       the event will be missed. When you attach a mailbox to an event,
+ *       make always sure to have enough space in the mailbox.
```

Comment:
Should we then disable the interrupts and query it with `kmbox_full` ?

Reply:
I think it's up to the receiver thread to create a mailbox which is large enough to buffer pending events that might be received. In general, I would say that most threads will wait only on 1 event (1 type of interrupt), and then process it as soon as it happens and before it happens again.

On the sender side, we're already under interrupt. So if a AI interrupt happens, the code will notify all pending threads that were waiting for an AI interrupt. But if one of those mailboxes happen to be full, there's nothing we can do in the context of the AI interrupt.